tools/gfx/cuda/render-cuda.cpp

```diff
@@ -954,13 +956,367 @@ class CUDARenderer : public RendererBase
+                cuModuleGetGlobal(
+                    &globalParamsSymbol,
+                    &globalParamsSymbolSize,
+                    currentPipeline->shaderProgram->cudaModule,
+                    "SLANG_globalParams");
+
+                CUdeviceptr globalParamsCUDAData =
+                    currentRootObject->bufferResource
+                        ? (CUdeviceptr)currentRootObject->bufferResource->getBindlessHandle()
+                        : 0;
+                cudaMemcpyAsync(
+                    (void*)globalParamsSymbol,
+                    (void*)globalParamsCUDAData,
+                    globalParamsSymbolSize,
+                    cudaMemcpyDeviceToDevice,
+                    0);
+            }
+            //
+            // The argument data for the entry-point parameters are already
+            // stored in host memory in a CUDAEntryPointShaderObject, as expected by cuLaunchKernel.
+            //
+            auto entryPointBuffer = currentRootObject->entryPointObjects[kernelId]->getBuffer();
+            auto entryPointDataSize =
+                currentRootObject->entryPointObjects[kernelId]->getBufferSize();
+
+            void* extraOptions[] = {
+                CU_LAUNCH_PARAM_BUFFER_POINTER,
+                entryPointBuffer,
+                CU_LAUNCH_PARAM_BUFFER_SIZE,
+                &entryPointDataSize,
+                CU_LAUNCH_PARAM_END,
+            };
+
+            // Once we have all the decessary data extracted and/or
+            // set up, we can launch the kernel and see what happens.
+            //
+            auto cudaLaunchResult = cuLaunchKernel(
+                currentPipeline->shaderProgram->cudaKernel,
+                x,
+                y,
+                z,
+                int(threadGroupSize[0]),
+                int(threadGroupSize[1]),
+                int(threadGroupSize[2]),
+                sharedSizeInBytes,
+                stream,
+                nullptr,
+                extraOptions);
+
+            SLANG_ASSERT(cudaLaunchResult == CUDA_SUCCESS);
+        }
+
+        void copyBuffer(
+            IBufferResource* dst,
+            size_t dstOffset,
+            IBufferResource* src,
+            size_t srcOffset,
+            size_t size)
+        {
+            auto dstImpl = static_cast<MemoryCUDAResource*>(dst);
+            auto srcImpl = static_cast<MemoryCUDAResource*>(src);
+            cudaMemcpy(
+                (uint8_t*)dstImpl->m_cudaMemory + dstOffset,
+                (uint8_t*)srcImpl->m_cudaMemory + srcOffset,
+                size,
+                cudaMemcpyDefault);
+        }
+
+        void uploadBufferData(IBufferResource* dst, size_t offset, size_t size, void* data)
+        {
+            auto dstImpl = static_cast<MemoryCUDAResource*>(dst);
+            cudaMemcpy((uint8_t*)dstImpl->m_cudaMemory + offset, data, size, cudaMemcpyDefault);
+        }
+
+        void execute(CommandBufferImpl* commandBuffer)
```

Comment:
It seems like the implementations that want to use user-space command buffers could all implement a common interface like `IImmediateModeCommands` and then we could use a single `execute()` implementation that can work with any of them rather than need to implement the walking of the command list multiple times.

Maybe that wouldn't be worth the cost, because then the CUDA renderer would need stub functions for all the commands that could conceivably be recorded but that it doesn't support.